server/queues/processors/WebsocketsProcessor.ts

```diff
@@ -466,153 +480,172 @@ export default class WebsocketsProcessor {
-              .emit("collections.remove_user", {
+              socketio
+                .to(`user-${event.userId}`)
+                .emit(
+                  "collections.add_group",
+                  presentGroupMembership(groupMembership)
+                );
+
+              // tell any user clients to connect to the websocket channel for the collection
+              socketio.to(`user-${event.userId}`).emit("join", {
                 event: event.name,
-                userId: event.userId,
                 collectionId: groupMembership.collectionId,
               });
-            // tell any user clients to disconnect from the websocket channel for the collection
-            socketio.to(`user-${event.userId}`).emit("leave", {
-              event: event.name,
-              collectionId: groupMembership.collectionId,
-            });
+            }
           }
+        );
+
+        return;
+      }
+
+      case "groups.remove_user": {
+        const membership = {
+          event: event.name,
+          userId: event.userId,
+          groupId: event.modelId,
+        };
+
+        // let everyone with access to the group know a user was removed
+        socketio
+          .to(`team-${event.teamId}`)
+          .emit("groups.remove_user", membership);
+
+        socketio.to(`user-${event.userId}`).emit("leave", {
+          event: event.name,
+          groupId: event.modelId,
+        });
+
+        const user = await User.findByPk(event.userId);
+        if (!user) {
+          return;
         }
 
+        await GroupMembership.findAllInBatches<GroupMembership>(
+          {
+            where: {
+              groupId: event.modelId,
+            },
+            limit: 100,
+          },
+          async (groupMemberships) => {
+            for (const groupMembership of groupMemberships) {
+              if (!groupMembership.collectionId) {
+                continue;
+              }
+
+              socketio
+                .to(`user-${event.userId}`)
+                .emit(
+                  "collections.remove_group",
+                  presentGroupMembership(groupMembership)
+                );
+
+              const collection = await Collection.scope({
+                method: ["withMembership", event.userId],
+              }).findByPk(groupMembership.collectionId);
+
+              if (cannot(user, "read", collection)) {
+                // tell any user clients to disconnect from the websocket channel for the collection
+                socketio.to(`user-${event.userId}`).emit("leave", {
+                  event: event.name,
+                  collectionId: groupMembership.collectionId,
+                });
+              }
+            }
+          }
+        );
+
         return;
       }
 
       case "groups.delete": {
         socketio.to(`team-${event.teamId}`).emit(event.name, {
           modelId: event.modelId,
         });
+        socketio.to(`group-${event.modelId}`).emit("leave", {
```

Comment:
@apoorv-mishra I'm confident we can remove all these `leave` events in the near future with an abstraction on the client that maps a model removal to a channel disconnect

Reply:
So I thought. Similar analogy holds for `join` events as well. I can take it up separately.

Reply:
Easy enough when the entire channel is disappearing, such as `groups.delete` but for kicking individual users out I still don't have a good method that doesn't involve a client roundtrip.